criu/cr-restore.c

```diff
@@ -1845,18 +1848,26 @@ static int restore_task_with_children(void *_arg)
 		/* Wait prepare_userns */
 		if (restore_finish_ns_stage(CR_STATE_ROOT_TASK, CR_STATE_PREPARE_NAMESPACES) < 0)
 			goto err;
+
+		parent_cg_set = 0;
 	}
 
 	if (needs_prep_creds(current) && (prepare_userns_creds()))
 		goto err;
 
+	/* Zombies and helpers can have cg_set == 0 so we skip them */
+	while (parent && !rsti(parent)->cg_set && parent->parent) {
```

Comment:
The original code didn't have the `&& parent->parent` check in the if condition. Why is it needed here?

Reply:
If parent->parent (the grandparent item, if you will) is NULL, we may get another segfault for calling `rsti(parent)->cg_set` after assigning parent->parent as the new parent.

Reply:
But it is functionally different from the existing code. Take this chain for example
a(cg_set = 0) <- b (cg_set = 0) <- c (cg_set=0) <- current
Your loop will exit with parent = a

But the original code would have exited the loop with 
parent = null

This makes the code materially different.

Reply:
Also, grandparent being NULL is valid. The problem is that we've changed the logic. In the original code, if parent is null, it takes root_cg_set. But with the example I just gave, it will take cg_set = 0.

Reply:
My intention with this change is so cg_set = 0 can make the following check in `prepare_task_cgroups` work similarly to how the original code intended:
```if (parent_cg_set)
		current_cgset = parent_cg_set;
	else
		current_cgset = root_cg_set;```
I see your point in how the added `parent->parent` condition may exclude a NULL grandparent, and I will modify the code so it's clearer -- however, I believe the code as is would exit the loop with parent = a, pass in parent_cg_set = 0 to `prepare_task_cgroups` and still assign current_cgset = root_cg_set.

Reply:
Although I think Angie is right and it would work fine, I agree with Kashyap that it would be easier/clearer if you just moved the whole initialization of `current_cgset` here.